_posts/2021-06-22-type-phantoms-and-safe-apis.md

```diff
@@ -0,0 +1,712 @@
+): Handlers<T> {
+  /* ... */
+}
+```
+
+Now the common utility functions don't know the details of how to make requests, and those details
+can be tweaked freely. Some features still need type changes. For example, using the `Result` type
+means changing the `MakeRequest` type as well.
+
+The server implementation could use the same principles to extract library logic and enable
+customization.
+
+### Simplifying types
+
+Making things simpler can make them easier to use as well. The utility types are quite complex, and
+perhaps some could even be removed entirely? It's easy to make some types simpler, but type
+inference can suffer, and the error messages could become more cryptic.
+
+It's also possible to derive all the API model types from the server-side implementation. Then
+separate types wouldn't have to be written at all. However, this would make the API model less
+understandable in most cases and make some error messages incomprehensible.
+
+Despite these challenges, there's probably a lot that could be improved, without adverse effects on
+readability.
+
+### Stop being so RESTful?
+
+If you are just writing function calls, do you need separate body, path and query parameters? Are
+the different HTTP methods useful?
+
+If it's an external API, sticking to typical REST conventions will be more familiar to the API
+users. For an internal API, the API structure may not matter that much when it's abstracted away.
+Ditching different parameter types would already make utilities simpler. Perhaps only GET and POST
+requests are needed? Maybe route paths are unnecessary too, and the function name could be passed as
+a parameter instead?
+
+GraphQL for example [serves data this way](https://graphql.org/learn/serving-over-http/). However,
+sticking with some REST conventions may be a better fit for a project, and custom utilities allow
+flexibility.
+
+When using common utilities for the API, it's easy to experiment with changes like this. Just
+modifying the utilities will change how all endpoints work. Validating and testing that an
+experiment works doesn't require a lot of changes.
+
+## Verdict: Definitely recommended
+
+What benefits does a type-safe API give us?
+
+A good data model can make many errors outright impossible. Shared, common types are a single source
+of truth for how the data should look. A straightforward API model enables more use cases, such as
+required validation.
+
+When some functionality changes, the types can tell what other parts of the code should change as
+well. While it would be ideal to place all related code in one location, that's not always feasible.
+API changes are a common example: Placing all related frontend and backend code in the same file
+isn't possible in most cases. In this situation the contract of types is the next best thing: We'll
+get a compiler error for any incompatible changes. Similarly, if we forget something when
+implementing a new feature or endpoint, the compiler will remind us.
```

Comment:
Definitely agree with this, and the verdict as a whole. Having lived with a system like this (that you know very well) in production for the past year and a half, I don't have complaints. Quite the opposite, actually. The API has been one of the safer places to work with, even when inducting developers unfamiliar with the system or TypeScript itself. I'd go as far as to call static typing a kind of "computer-aided design".